### ISSUE TITLE
predict with missing values

### ISSUE THREAD

**Issue by jlevatic** [imbs-hl/ranger#94]
Why ranger can't make predictions when there are missing values in the data? In predict.R file there are the following lines (160-162):

`function return :`if (any(is.na(data.final))) {
    stop("Missing values in data.")
  }`

Ranger successfully trains random forest on data which contains missing values, so it makes me think why it can't make predictions as well?

**Comment by mayer79** [comment]
What version are you using? At least the current CRAN release (0.5.0) would not allow missing values in the training data. But certainly we would all be extremely happy if `ranger` would be able to treat NA ;).

**Comment by jlevatic** [comment]
I'm using version 0.4.2. Ranger did not complain about missing values in the training data, however, predict function did. 

If I may ask, what is hindering ranger from handling NAs? Tree based methods can handle NAs nicely, i.e.,  an example with missing value is usually passed to all of the branches of the tree, or alternatively, in random forest you can chose one branch at random, it usually works equally well as the former approach and it's faster.

**Comment by mnwright** [comment]
In training, the data is checked for missing data if the formula interface is used. If the alternative interface (dependent.variable.name) is used, this check is currently skipped. I'm not sure what's exactly happening with the NA's in this case, so be careful. 
We should check the runtime of this check for huge datasets and probably add it (if possible with column information, as in #89). 

In prediction, ranger always checks for missing values. 

@jlevatic you are right, there are some options to handle NA's in random forests. We are still not sure what's the best option in which case and therefore there is nothing implemented yet. But I have a student working on that. ;)

**Comment by mnwright** [comment]
Better NA checks now included in #109.

**Comment by nlapidot3** [comment]
@mnwright do you have any updates on implementing missing data imputation within ranger?

**Comment by mnwright** [comment]
Unfortunately not. We achieved good results with multiple imputation and adaptive tree imputation (Ishwaran 2008, http://dx.doi.org/10.1214/08-AOAS169). However, there is nothing implemented in ranger yet.

**Comment by NamLQ** [comment]
@nlapidot3 
Maybe you need this https://github.com/mayer79/missRanger

**Comment by nlapidot3** [comment]
@mnwright thanks for the reference.
@NamLQ thank you for the suggestion. I will take a look at it and see if it works for my purposes.

**Comment by mayer79** [comment]
`missRanger` is built to impute missing values in a data set. To use such chaining loops within `ranger` would be awfully inefficient. There *must* be much much better ways to deal with missing predictor values in a random forest! One thing that could work with few missings: if an observation with missing x is to be split on x, send part of its case weight to the left and the remaining weight to the right. Same during prediction.

**Comment by samFarrellDay** [comment]
I'm coming to this thread because I'm actually working on a homebrew package that uses ranger to impute missing values. Ranger is the fastest rf package I know of so I was hoping I could bootleg the mice package and use ranger instead of randomForest.

I'm sure most of you know this but xgboost handles missing variables by putting them in the split that minimizes the loss function the most, which I think is an elegant solution.

**Comment by mnwright** [comment]
In my understanding you don't need to predict with missing values in multiple imputation. 

> I'm sure most of you know this but xgboost handles missing variables by putting them in the split that minimizes the loss function the most, which I think is an elegant solution.

I think that is a very promising solution. Still, I think we need a good comparison before we decide on missing value methods to implement in ranger,

**Comment by Geoff-Kahn** [comment]
Just wondering if there has been any progress in implementing a method for missing values? The suggestion above of putting them in the split that minimizes loss seems like a good one.

**Comment by mnwright** [comment]
Sorry, no progress.
